File: plugins/codex-forge/scripts/codegraph.py
#!/usr/bin/env python3
from __future__ import annotations

import os
import shutil
import sys

PACKAGE = "@colbymchenry/codegraph"


def command_candidates() -> list[list[str]]:
    candidates: list[list[str]] = []
    if shutil.which("codegraph"):
        candidates.append(["codegraph"])
    if shutil.which("bun"):
        candidates.append(["bun", "x", PACKAGE])
    elif shutil.which("bunx"):
        candidates.append(["bunx", PACKAGE])
    if shutil.which("pnpm"):
        candidates.append(["pnpm", "dlx", PACKAGE])
    elif shutil.which("pnpx"):
        candidates.append(["pnpx", PACKAGE])
    if shutil.which("yarn"):
        candidates.append(["yarn", "dlx", PACKAGE])
    if shutil.which("npx"):
        candidates.append(["npx", "--yes", PACKAGE])
    return candidates


def resolve_command() -> list[str] | None:
    candidates = command_candidates()
    return candidates[0] if candidates else None


def main() -> int:
    command = resolve_command()
    if command is None:
        print(
            "CodeGraph CLI unavailable: install Bun, pnpm, Yarn, or npm/Node.js",
            file=sys.stderr,
        )
        return 127
    os.execvp(command[0], [*command, *sys.argv[1:]])
    return 127


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/codex-forge/scripts/codegraph.mjs
#!/usr/bin/env bun
import { spawnSync } from "node:child_process";
import { which } from "./lib/process.mjs";

export const PACKAGE = "@colbymchenry/codegraph";

export async function commandCandidates(resolve = which) {
	const candidates = [];
	if (await resolve("codegraph")) candidates.push(["codegraph"]);
	if (await resolve("bun")) candidates.push(["bun", "x", PACKAGE]);
	else if (await resolve("bunx")) candidates.push(["bunx", PACKAGE]);
	if (await resolve("pnpm")) candidates.push(["pnpm", "dlx", PACKAGE]);
	else if (await resolve("pnpx")) candidates.push(["pnpx", PACKAGE]);
	if (await resolve("yarn")) candidates.push(["yarn", "dlx", PACKAGE]);
	if (await resolve("npx")) candidates.push(["npx", "--yes", PACKAGE]);
	return candidates;
}

export async function resolveCommand(resolve = which) {
	return (await commandCandidates(resolve))[0] ?? null;
}

async function main() {
	const command = await resolveCommand();
	if (!command) {
		process.stderr.write(
			"CodeGraph CLI unavailable: install Bun, pnpm, Yarn, or npm/Node.js\n",
		);
		return 127;
	}
	const result = spawnSync(
		command[0],
		[...command.slice(1), ...process.argv.slice(2)],
		{ stdio: "inherit" },
	);
	return result.status ?? 127;
}

if (import.meta.main) process.exit(await main());
